### PR TITLE
return error when something went wrong instead of panic

### PR DESCRIPTION
**Pull request by amirvalhalla**

### CODE REVIEW

File: queue.go [hunk 1]
```diff
@@ -7,14 +7,26 @@ The queue implemented here is as fast as it is for an additional reason: it is *
 */
 package queue
 
+import "sync"
+
+import (
```
Comment (lil5) [inline queue.go:10]: Use go fmt it'll merge the imports into one brackets list

File: queue_test.go [hunk 2]
```diff
@@ -26,15 +26,15 @@ func TestQueueWrapping(t *testing.T) {
...
 	}
 
 	for i := 0; i < minQueueLen; i++ {
-		if q.Peek().(int) != i+3 {
-			t.Error("peek", i, "had value", q.Peek())
+		if r, _ := q.Peek(); r.(int) != i+3 {
+			t.Error("peek", i, "had value", r)
 		}
-		q.Remove()
+		_, _ = q.Remove()
```
Comment (lil5) [inline queue_test.go]: Unnecessary

File: queue_test.go [hunk 3]
```diff
@@ -52,7 +52,7 @@ func TestQueueLength(t *testing.T) {
 		}
 	}
 	for i := 0; i < 1000; i++ {
-		q.Remove()
+		_, _ = q.Remove()
```
Comment (lil5) [inline queue_test.go]: Unnecessary

**Review by lil5 — CHANGES_REQUESTED** [review]

**Comment by eapache** [comment]
Similar to https://github.com/eapache/queue/pull/24#issuecomment-1501176762, the use of `panic` in this library was intentional - it mimics as close as possible a low-level built-in structure, where a failure is exceptional, the result of programmer error (so `panic` is correct) and not a normal operating condition (where returning errors would be correct).

As with the other ticket, I’m willing to consider a secondary package with this behaviour instead, but even then I would question whether returning user-space errors from this package is ever correct?

**Comment by amirvalhalla** [comment]
> 

in some cases it could be true , e.g. the user will dequeue as long as it reach to empty error without checking length of queue or in similar cases

**Comment by eapache** [comment]
It takes just as much code for the client to check the queue length before as it does for them to handle the error after (probably less, because the length check can be mixed into the loop condition).

**Pull request closed without merging**